### DISCUSSION TITLE
자바스크립트의 모듈 시스템에는 무엇이 있나요?

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#102] [모듈]
- 자바스크립트의 모듈 시스템에는 무엇이 있나요?
- 각 모듈의 특징을 설명해주세요.
  - 모듈 시스템의 차이점을 설명해주세요.

**Comment by publdaze** [comment]
> * 자바스크립트의 모듈 시스템에는 무엇이 있나요?
- ES6 모듈, CommonJS, AMD
> * 각 모듈의 특징을 설명해주세요.
>   * 모듈 시스템의 차이점을 설명해주세요.
- ES6 모듈 - import/export 사용, 항상 strict mode 적용
- CommonJS - require/module.exports,exports 사용, 동기 작업, 주로 서버에서 사용, Node.js에서 CommonJS 사용한 독자적 모듈 환경 제공
- AMD - require/define 사용, 비동기 작업, 주로 클라이언트에서 사용

**Comment by pipisebastian** [comment]
> * 자바스크립트의 모듈 시스템에는 무엇이 있나요?
ES6 모듈, CommonJS, AMD
> * 각 모듈의 특징을 설명해주세요.
>   * 모듈 시스템의 차이점을 설명해주세요.
- **CommonJS**
    - 동기적인 특성 → 서버 사이드에서 용이
    - NodeJs에서 사용하는 방식 (독자적으로 진화시켰지만, 기본적으로 CommonJs를 따른다)
    - 불러오기 `require()`, 내보내기 `module.exports` `exports`
- **ES6 모듈 (ESM)**
    - 비동기적인 특성으로, 브라우저에서 효율적임.
    - 불러오기 `import`, 내보내기 `export`
    - 브라우저에서 -> `script` 태그에 `type = “module”` 어트리뷰트 추가
    - ESM 임을 명확하게 하기 위해, **확장자 mjs** 사용할 것 권장
    - ESM에서는 기본적으로 `strict mode` 적용
- **AMD**
    - 비동기적인 특성으로, 브라우저에서 효율적임.
    - 불러오기 `require`, 내보내기 `define`
